recipes/lora_finetune_distributed_td.py

```diff
@@ -0,0 +1,1119 @@
+        num_training_steps: int,
+        last_epoch: int,
+    ) -> Optimizer:
+        lr_scheduler = config.instantiate(
+            cfg_lr_scheduler,
+            self._optimizer,
+            num_training_steps=num_training_steps,
+            last_epoch=last_epoch,
+        )
+        if self._is_rank_zero:
+            log.info("Learning rate scheduler is initialized.")
+        return lr_scheduler
+
+    def _setup_one_dataset(
+        self,
+        cfg_dataset: DictConfig,
+        global_shuffle: bool,
+        global_parallel_method: str,
+        global_streaming: bool,
+        global_num_workers: int,
+    ):
+        streaming = cfg_dataset.pop("streaming", global_streaming)
+        parallel_method = cfg_dataset.pop("parallel_method", global_parallel_method)
+        shuffle = cfg_dataset.pop("shuffle", global_shuffle)
+        num_workers = cfg_dataset.pop("num_workers", global_num_workers)
+
+        ds = config.instantiate(
+            cfg_dataset,
+            self._tokenizer,
+            streaming=streaming,
+            shuffle=shuffle,
+            parallel_method=parallel_method,
+            num_workers=num_workers,
+        )
+
+        return ds
+
+    def _setup_data_td(
+        self,
+        cfg_dataloader: DictConfig,
+        batch_size: int,
+    ) -> BaseNode:
+        """
+        All data related setup happens here. Currently this recipe only supports
+        Map-style Datasets which fit into memory and an option for random shuffling.
+        Samplers, iterable datasets, and streaming datasets are not supported.
+        """
+        world_size, rank = training.get_world_size_and_rank()
+
+        # Get global settings
+        shuffle = cfg_dataloader.shuffle
+        parallel_method = cfg_dataloader.get("parallel_method", "thread")
+        packed = cfg_dataloader.get("packed", False)
+        streaming = cfg_dataloader.get("streaming", False)
+        num_workers = cfg_dataloader.get("num_workers", 0)
+        pin_memory = cfg_dataloader.get("pin_memory", True)
+        collate_fn = cfg_dataloader.collate_fn
+
+        cfg_datasets = cfg_dataloader["datasets"]
+        weights, datasets = {}, {}
+        if isinstance(cfg_datasets, DictConfig):
+            for k, cfg_and_weight in cfg_datasets.items():
+                weights[k] = float(cfg_and_weight.get("weight", 1.0))
+                datasets[k] = Prefetcher(
+                    self._setup_one_dataset(
+                        cfg_dataset=cfg_and_weight.dataset,
+                        global_shuffle=shuffle,
+                        global_parallel_method=parallel_method,
+                        global_streaming=streaming,
+                        global_num_workers=num_workers,
+                    ),
+                    prefetch_factor=8,
+                )
+
+        # Instantiate collate_fn
+        if "left_pad_sequence" in collate_fn:
+            raise RuntimeError("left_pad_sequence collator is only for inference.")
```

Comment:
I understand that this was already in the setup_data, but we usually try to fail fast and catch errors like this in the init.